Use filter files

r-pascua opened the pull request:
This PR adds the ability to use the tophat filtering code with `yaml` files containing pre-computed filter parameters. The core assumption here is that there is one filter file per spectral window, and the filter files have the following structure:
```
filter_centers:
    (ant1, ant2): blah
filter_half_widths:
    (ant1, ant2): blah
```
where `(ant1, ant2)` here is a placeholder for all of the baselines in the array (or a superset of baselines in the array). I am open to changing the file format slightly, since `yaml.SafeLoader` doesn't know how to interpret python tuples; we would just need to agree on a format to use, document it, and stick with it. I have written simple unit tests for the code functionality, but I haven't yet tried a more sophisticated test (i.e., on something that looks like real data with many baselines).

I also added some line breaks here and there because the code was wrapping in my terminal.

jsdillon reviewed (changes requested):
Looks pretty good. Just a handful of minor issues, mostly in documentation and explanations.

For future reference, I'd really prefer if unrelated whitespace changes are put in a separate PR. Putting it all together makes it a lot more time consuming to review because I have to carefully scan for the substantive changes.

r-pascua commented on hera_cal/frf.py:
```diff
@@ -1672,143 +1679,245 @@ def load_tophat_frfilter_and_write(datafile_list, case, baseline_list=None, calf
+            have_bl_info.append(
+                (bl[:2] in filter_antpairs) or (bl[:2][::-1] in filter_antpairs)
+            )
+        if not all(have_bl_info):
```
How about I update this so that when it errors it tells the user all of the baselines that couldn't be found?

jsdillon replied:
that's fine

r-pascua commented on hera_cal/frf.py:
```diff
@@ -1672,143 +1679,245 @@ def load_tophat_frfilter_and_write(datafile_list, case, baseline_list=None, calf
-                assert case in ['sky', 'max_frate_coeffs', 'uvbeam'], f'case={case} is not valid.'
-                # use conservative nfr (lowest resolution set).
+                # Otherwise, we need to compute the filter parameters.
+                if case not in ("sky", "max_frate_coeffs", "uvbeam"):
```
Yeah, I can add a new option for `case` and update the documentation.

r-pascua commented on hera_cal/tests/test_frf.py:
```diff
@@ -930,6 +931,92 @@ def test_load_dayenu_filter_and_write(self, tmpdir):
+        uvh5 = os.path.join(
+            DATA_PATH, "test_input/zen.2458101.46106.xx.HH.OCR_53x_54x_only.uvh5"
+        )
+        frate_centers = {(53, 54): 0.1}
+        frate_half_widths = {(53, 54): 0.05}
+        filter_info = dict(
+            filter_centers=frate_centers,
+            filter_half_widths=frate_half_widths,
+        )
+        frate_centers = {k+("ee",): v for k, v in frate_centers.items()}
+        frate_half_widths = {k+("ee",): v for k, v in frate_half_widths.items()}
+
+        with open(tmpdir / "filter_info.yaml", "w") as f:
+            yaml.dump(filter_info, f)
```
Sounds good! I'd like to come to a consensus on how to format the yaml files before including an example file, though.

jsdillon replied:
Are there any real decisions to be made? I'm fine with you defining the format, as long as you think it's general enough for any use cases we might consider.

I guess the main questions are: 1) how to handle the case where the use might want to handle multiple polarizations (or no polarization, so same filter parameters for all pols) and 2) how to handle different baseline keys for redundant baselines

r-pascua replied:
We chatted about some of this via slack, but I figure I should echo things here so we have a public record. 

1) The filtering code assumes that, for a given baseline and spectral window, all polarizations use the same fringe-rate filter.
2) I think I would like to enforce that filter parameter files contain at least all of the baselines present in the data. I suppose that it is possible to use some of the abscal tools to figure out how to do the mapping from a smaller set of baselines to those in the data, so I can add that feature if you insist on having it.

jsdillon replied:
Let's leave it out for now, but we may want to revisit in the future.

r-pascua commented on hera_cal/frf.py:
```diff
@@ -1572,24 +1574,29 @@ def tophat_frfilter_argparser(mode='clean'):
             "This does not change the format of the output files but it does change the nature of their content.")
     filt_options.add_argument("--ninterleave", default=1, type=int, help=desc)
+    filt_options.add_argument(
+        "--param_file", default="", type=str, help="File containing filter parameters"
```
Yeah, definitely. I kept it terse because I wanted to wait until we agreed on how to format the files.

r-pascua commented:
Thank you for the review @jsdillon! I will make the recommended changes. I apologize for adding a bunch of fluff in the form of whitespace changes, but I develop in vim and it was practically impossible to read the code in the terminal with how much line wrapping there was. I will keep this in mind if I make future additions to the code.

Before addressing your comments, I would like to get your opinion on the file format. Ideally, I would access the filter parameters with an antenna pair tuple; however, python tuples are not supported by `yaml.SafeLoader` (which is what I figure we would want to use). I have two proposals, and would like your input on which you would prefer.

1. Use baseline integers for the dictionary keys, assuming 350 antennas. This would require an extra step of converting the baseline integers to antenna pair tuples after reading in the file, but this feels like a very easy conversion to the format expected by the filtering code.
2. Use strings for the dictionary keys, formatted like `"(0,1)"` for example. This feels like the worse of the two options, since this is susceptible to whitespace issues (e.g., assuming this format would cause the code to miss keys formatted like `"(0, 1)"`), and generally feels kind of clunky.

Thank you again for the review! I can make all of the requested changes once we settle on a final format for the filter files.

jsdillon commented:
What do you mean by baseline integers? Like the ones in pyuvdata? Those aren't human readable...

r-pascua commented:
Yeah, like the ones from `pyuvdata`. The main problem I'm butting up against is that `yaml.SafeLoader` doesn't work with python tuples as keys, and using tuples as keys gives really wonky looking yaml files. Here's an example:
```
filter_centers:
  ? &id001 !!python/tuple
  - 0
  - 1
  : 0.123
  ? &id002 !!python/tuple
  - 0
  - 2
  : 0.173
filter_half_widths:
  *id001: 0.07
  *id002: 0.08
```

r-pascua commented:
OK, I have updated the code so that filter parameter files use strings for the antenna pairs, and there is an extra bit of code that converts those strings into tuples. I have also fleshed out the documentation regarding the parameter files and added an example parameter file. Here is what the sample file looks like:

```
filter_centers:
  (0, 1): 0.1234
  (0, 2): 0.173
filter_half_widths:
  (0, 1): 0.05
  (0, 2): 0.08
```

Please let me know what additional changes you would like to see when you have time to go through the changes.

jsdillon commented on hera_cal/tests/test_frf.py:
```diff
@@ -930,6 +931,97 @@ def test_load_dayenu_filter_and_write(self, tmpdir):
+        frate_centers = {k+("ee",): v for k, v in frate_centers.items()}
+        frate_half_widths = {k+("ee",): v for k, v in frate_half_widths.items()}
+
+        with open(tmpdir / "filter_info.yaml", "w") as f:
+            yaml.dump(filter_info, f)
```
One final suggestion: let's use the exmaple_filter_params.yaml in the test (that ensures it's not accidentally deleted in the future)

r-pascua replied:
Hm. How about I use the example filter file in the test for the missing baseline error message? The one that tests the actual performance of the code has two cases to make sure it does things correctly when the antennas are flipped in the filter parameter file, so I'm partial to using a temporary file for those tests.

jsdillon replied:
That's fine

jsdillon reviewed (approved):
One last request, but after that, good to go!
